File: transfer/Int2Numeral.hs
module Int2Numeral (int2numeralTree, int2numeralInTree) where

import PGF

-- to test:
--    echo "pg -pgf" | gf -run alltenses/NumeralEng.gfo
-- or some other language, then
--    runghc Int2Numeral
-- which waits for numeral input.
-- If you just press <enter>, a predefined list of samples is converted.
--
-- to use in your own program: see Int2NumeralExample.hs


int2numeralInTree :: [(CId, CId)] -> Tree -> Tree
int2numeralInTree idmap t = case unApp t of
  Just (c,ts) -> case lookup c idmap of
    Just f | length ts == 1 -> mkApp f [int2numeralTree (head ts)]  
    _ -> mkApp c (map (int2numeralInTree idmap) ts)
  _ -> t

int2numeralTree :: Tree -> Tree
int2numeralTree = num . whole . intn . show . getInt
 where
  intn ds = case ds of
    _ | length ds > 6 -> error $ "sorry, " ++ ds ++ " too large for int2numeral"
    '0':dd  -> intn dd 
    "1"     -> (pot2as3 , pot1as2 , pot0as1 , pot01)
    [d]     -> (pot2as3 , pot1as2 , pot0as1 , pot0 (nd d))
    "10"    -> (pot2as3 , pot1as2 , id      , pot110)
    "11"    -> (pot2as3 , pot1as2 , id      , pot111)
    ['1',d] -> (pot2as3 , pot1as2 , id      , pot1to19 (nd d))
    [d,'0'] -> (pot2as3 , pot1as2 , id      , pot1 (nd d))
    [d,e]   -> (pot2as3 , pot1as2 , id      , pot1plus (nd d) (unit (intn [e])))
    [d,'0','0'] -> (pot2as3 , id  , id      , pot2 (unit (intn [d])))
    [d, e, f]   -> (pot2as3 , id  , id      , pot2plus (unit (intn [d])) (minor (intn [e,f])))
    _ -> case splitAt (length ds - 3) ds of
      (n,"000") -> (id, id, id, pot3 (major (intn n)))
      (n,m)     -> (id, id, id, pot3plus (major (intn n)) (major (intn m)))
  whole (f,g,h,t) = f (g (h t))
  major (_,g,h,t) = g (h t)
  minor (_,g,h,t) = h t
  unit  (_,_,_,t) = t
  getInt t = case unInt t of
    Just n -> n
    _ -> error $ "cannot get Int from " ++ showExpr [] t

-- auxiliaries

app :: String -> [Tree] -> Tree
app f ts = mkApp (mkCId f) ts

atom :: String -> Tree
atom c = app c []

num t = app "num" [t]
pot2as3 t = app "pot2as3" [t]
pot1as2 t = app "pot1as2" [t]
pot0as1 t = app "pot0as1" [t]
pot3 t = app "pot3" [t]
pot2 t = app "pot2" [t]
pot1 t = app "pot1" [t]
pot0 t = app "pot0" [t]
pot1to19 t = app "pot1to19" [t]
pot1plus t u = app "pot1plus" [t,u]
pot2plus t u = app "pot2plus" [t,u]
pot3plus t u = app "pot3plus" [t,u]

pot110 = atom "pot110"
pot111 = atom "pot111"
pot01 = atom "pot01"
nd d = atom ("n" ++ [d])

-- test function

test = do
  pgf <- readPGF("Numeral.pgf")
  let eng:_ = languages pgf
  s <- getLine
  let ns = if null s then [1,2,8,10,11,20,21,40,95,100,101,234,1001,2021,630511] else [read s] 
  flip mapM_ ns $ \n -> do
    let t = int2numeralTree (mkInt n)
    putStrLn $ unwords [show n, linearize pgf eng (t)]

main = test

File: transfer/int2numeral.py
import pgf

"""
to test:
    echo "pg -pgf" | gf -run alltenses/NumeralEng.gfo
or some other language, then
    python3 int2numeral.py
which waits for numeral input.
If you just press <enter>, a predefined list of samples is converted.
    
to use in your own program: see int2numeral_example.py
"""

def int2numeral_parts(ds):
    if not (0 < len(ds) < 7):
        print("input length not in range 1--6:", ds)
    elif ds[0] == '0':
        return int2numeral_parts(ds[1:])
    elif ds == '1':
        whole, major, minor, unit = "pot2as3", "pot1as2", "pot0as1", const("pot01")
    elif len(ds) == 1:
        whole, major, minor, unit = "pot2as3", "pot1as2", "pot0as1", app("pot0",[nd(ds)])
    elif ds == '10':
        whole, major, minor, unit = "pot2as3", "pot1as2", "id",      const("pot110")
    elif ds == '11':
        whole, major, minor, unit = "pot2as3", "pot1as2", "id",      const("pot111")
    elif len(ds) == 2 and ds[0] == '1':
        whole, major, minor, unit = "pot2as3", "pot1as2", "id",      app("pot1to19",[nd(ds[1])])
    elif len(ds) == 2 and ds[1] == '0':
        whole, major, minor, unit = "pot2as3", "pot1as2", "id",      app("pot1",[nd(ds[0])])
    elif len(ds) == 2:
        _,_,_,unit = int2numeral_parts(ds[1])
        whole, major, minor, unit = "pot2as3", "pot1as2", "id",      app("pot1plus",[nd(ds[0]),unit])
    elif len(ds) == 3 and ds[1:] == '00':
        _,_,_,unit = int2numeral_parts(ds[0])
        whole, major, minor, unit = "pot2as3", "id",      "id",      app("pot2", [unit])
    elif len(ds) == 3:
        _,_,_,unit1 = int2numeral_parts(ds[0])
        _,_,minor2,unit2 = int2numeral_parts(ds[1:])
        whole, major, minor, unit = "pot2as3", "id",      "id",      app("pot2plus", [unit1, app(minor2,[unit2])])
    else:
        fst,snd = ds[:len(ds)-3],ds[len(ds)-3:]
        if snd == '000':
            _,major,minor,unit = int2numeral_parts(fst)
            whole, major, minor, unit = "id", "id", "id", apps(["pot3",major,minor],unit)
        else:
            _,major1,minor1,unit1 = int2numeral_parts(fst)
            _,major2,minor2,unit2 = int2numeral_parts(snd)
            part1 = apps([major1,minor1], unit1)
            part2 = apps([major2,minor2], unit2)
            whole, major, minor, unit = "id","id", "id", app("pot3plus",[part1,part2])

    return whole, major, minor,unit

    
def int2numeral_tree(tree):
    up = tree.unpack()
    if type(up) == int:
        ds = str(up)
    whole,major,minor,unit = int2numeral_parts(ds)
    return apps(["num",whole,major,minor],unit)

    
def int2numeral_in_tree(int_numeral_dict, tree):
    fun, args = tree.unpack()
    if fun in int_numeral_dict and len(args) == 1:
        f_numeral = int_numeral_dict[fun]
        return pgf.Expr(f_numeral, [int2numeral_tree(args[0])])
    else:
        return pgf.Expr(fun, [int2numeral_in_tree(int_numeral_dict, t) for t in args])


############ auxiliaries #####################

def app(f,xs):
    if f=="id":
        return xs[0]
    else:
        return pgf.Expr(f,xs)

def apps(fs,x):
    fs.reverse()
    def rapps(hs):
        exp = x
        for h in hs:
            exp = app(h,[exp])
        return exp
    return rapps(fs)


def const(f):
    return app(f,[])

def nd(d):
    return const("n"+d)


############ test #####################

def main():
    gr = pgf.readPGF("Numeral.pgf")
    eng = list(gr.languages.values())[0]
    n = input()
    if n:
        ns = [int(n)]
    else:
        ns = [1,2,8,10,11,20,21,40,95,100,101,234,1000,1001,2021,630511]
    for n in ns: 
        print(n, eng.linearize(int2numeral_tree(pgf.readExpr(str(n)))))


if __name__ == "__main__":
    main()


